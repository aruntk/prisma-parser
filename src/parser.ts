import * as pegjs from 'pegjs'
import grammar from './grammar'
import { IPrismaAST } from './interface'
import printer from './printer'

const parser = pegjs.generate(grammar)
export default {
  parse: (prismaQL: string): IPrismaAST => parser.parse(prismaQL),
  print: printer,
}
