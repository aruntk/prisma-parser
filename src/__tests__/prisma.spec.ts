import parser from '../parser'

describe('PrismaParser', () => {
  it('Parses model and primitive type columns correctly', async () => {
    const schema = `
    model User {
      id     Int @id
      name   String
    }`
    const tree = parser.print(parser.parse(schema))
    expect(tree).toMatchSnapshot()
  })
  it('Identify reference type column correctly', async () => {
    const schema = `
    model User {
      id     Int @id
      posts  Post
    }

    model Post {
      id       Int @id
      author   User
    }
    `
    const tree = parser.print(parser.parse(schema))
    expect(tree).toMatchSnapshot()
  })
  it('Identify multiple columns', async () => {
    const schema = `
    model User {
      id     Int @id
      name   String[]
    }
    `
    const tree = parser.print(parser.parse(schema))
    expect(tree).toMatchSnapshot()
  })

  it('Identify optional columns', async () => {
    const schema = `
    model User {
      id     Int @id
      name   String?
    }
    `
    const tree = parser.print(parser.parse(schema))
    expect(tree).toMatchSnapshot()
  })

  it('Throws error if reference type is undefined', async () => {
    const schema = `
    model User {
      id     Int @id
      posts  Post[]
    }
    `
    expect(() => parser.parse(schema)).toThrowError('Data type Post is undefined')
  })
})
