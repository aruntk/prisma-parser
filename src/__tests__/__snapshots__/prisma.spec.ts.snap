// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PrismaParser Identifies multiple columns 1`] = `"{\\"type\\":\\"PrismaQL\\",\\"datasources\\":[],\\"generators\\":[],\\"models\\":[{\\"type\\":\\"model\\",\\"name\\":\\"User\\",\\"columns\\":[{\\"type\\":\\"Column\\",\\"name\\":\\"id\\",\\"data_type\\":{\\"name\\":\\"Int\\",\\"type\\":\\"PrimitiveType\\"},\\"optional\\":false,\\"multiple\\":false,\\"primaryKey\\":true},{\\"type\\":\\"Column\\",\\"name\\":\\"name\\",\\"data_type\\":{\\"name\\":\\"String\\",\\"type\\":\\"PrimitiveType\\"},\\"optional\\":false,\\"multiple\\":true}]}]}"`;

exports[`PrismaParser Identifies optional columns 1`] = `"{\\"type\\":\\"PrismaQL\\",\\"datasources\\":[],\\"generators\\":[],\\"models\\":[{\\"type\\":\\"model\\",\\"name\\":\\"User\\",\\"columns\\":[{\\"type\\":\\"Column\\",\\"name\\":\\"id\\",\\"data_type\\":{\\"name\\":\\"Int\\",\\"type\\":\\"PrimitiveType\\"},\\"optional\\":false,\\"multiple\\":false,\\"primaryKey\\":true},{\\"type\\":\\"Column\\",\\"name\\":\\"name\\",\\"data_type\\":{\\"name\\":\\"String\\",\\"type\\":\\"PrimitiveType\\"},\\"optional\\":true,\\"multiple\\":false}]}]}"`;

exports[`PrismaParser Identifies reference type column correctly 1`] = `"{\\"type\\":\\"PrismaQL\\",\\"datasources\\":[],\\"generators\\":[],\\"models\\":[{\\"type\\":\\"model\\",\\"name\\":\\"User\\",\\"columns\\":[{\\"type\\":\\"Column\\",\\"name\\":\\"id\\",\\"data_type\\":{\\"name\\":\\"Int\\",\\"type\\":\\"PrimitiveType\\"},\\"optional\\":false,\\"multiple\\":false,\\"primaryKey\\":true},{\\"type\\":\\"Column\\",\\"name\\":\\"posts\\",\\"data_type\\":{\\"name\\":\\"Post\\",\\"type\\":\\"ReferenceType\\"},\\"optional\\":false,\\"multiple\\":false}]},{\\"type\\":\\"model\\",\\"name\\":\\"Post\\",\\"columns\\":[{\\"type\\":\\"Column\\",\\"name\\":\\"id\\",\\"data_type\\":{\\"name\\":\\"Int\\",\\"type\\":\\"PrimitiveType\\"},\\"optional\\":false,\\"multiple\\":false,\\"primaryKey\\":true},{\\"type\\":\\"Column\\",\\"name\\":\\"author\\",\\"data_type\\":{\\"name\\":\\"User\\",\\"type\\":\\"ReferenceType\\"},\\"optional\\":false,\\"multiple\\":false}]}]}"`;

exports[`PrismaParser Parses datasource correctly 1`] = `"{\\"type\\":\\"PrismaQL\\",\\"datasources\\":[{\\"name\\":\\"db3\\",\\"type\\":\\"datasource\\",\\"declarations\\":[{\\"name\\":\\"provider\\",\\"init\\":{\\"type\\":\\"Literal\\",\\"value\\":\\"sqlite\\",\\"raw\\":\\"\\\\\\"sqlite\\\\\\"\\"}},{\\"name\\":\\"url\\",\\"init\\":{\\"type\\":\\"Literal\\",\\"value\\":\\"file:dev.db\\",\\"raw\\":\\"\\\\\\"file:dev.db\\\\\\"\\"}},{\\"name\\":\\"enabled\\",\\"init\\":{\\"type\\":\\"Literal\\",\\"value\\":true,\\"raw\\":\\"true\\"}}]}],\\"generators\\":[],\\"models\\":[]}"`;

exports[`PrismaParser Parses generator correctly 1`] = `"{\\"type\\":\\"PrismaQL\\",\\"datasources\\":[],\\"generators\\":[{\\"name\\":\\"photon\\",\\"type\\":\\"generator\\",\\"declarations\\":[{\\"name\\":\\"provider\\",\\"init\\":{\\"type\\":\\"Literal\\",\\"value\\":\\"photonjs\\",\\"raw\\":\\"\\\\\\"photonjs\\\\\\"\\"}}]}],\\"models\\":[]}"`;

exports[`PrismaParser Parses model and primitive type columns correctly 1`] = `"{\\"type\\":\\"PrismaQL\\",\\"datasources\\":[],\\"generators\\":[],\\"models\\":[{\\"type\\":\\"model\\",\\"name\\":\\"User\\",\\"columns\\":[{\\"type\\":\\"Column\\",\\"name\\":\\"id\\",\\"data_type\\":{\\"name\\":\\"Int\\",\\"type\\":\\"PrimitiveType\\"},\\"optional\\":false,\\"multiple\\":false,\\"primaryKey\\":true},{\\"type\\":\\"Column\\",\\"name\\":\\"name\\",\\"data_type\\":{\\"name\\":\\"String\\",\\"type\\":\\"PrimitiveType\\"},\\"optional\\":false,\\"multiple\\":false}]}]}"`;
